File: bindings/zig/src/safe_email.zig
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>
//
// Proven SafeEmail - FFI bindings to libproven email validation.
// All computation is performed in verified Idris 2 code via libproven.

const std = @import("std");
const c = @cImport(@cInclude("proven.h"));

/// Validate email address (RFC 5321 simplified) via libproven.
pub fn isValid(email: []const u8) bool {
    const result = c.proven_email_is_valid(email.ptr, email.len);
    return result.status == c.PROVEN_OK and result.value;
}

test "isValid" {
    try std.testing.expect(isValid("user@example.com"));
    try std.testing.expect(!isValid("not-an-email"));
    try std.testing.expect(!isValid("@invalid.com"));
}

File: bindings/rust/src/safe_email.rs
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>

//! Safe email validation via libproven FFI.
//!
//! Validates email addresses according to RFC 5321/5322 simplified rules.
//! All operations delegate to Idris 2 verified code.

use crate::core::{self, Result};
use crate::ffi;

/// Safe email validation.
pub struct SafeEmail;

impl SafeEmail {
    /// Validate an email address.
    ///
    /// Returns `true` if the email address is valid per RFC 5321 simplified rules.
    pub fn is_valid(email: &str) -> Result<bool> {
        let bytes = email.as_bytes();
        // SAFETY: We pass a valid pointer and length from the string's bytes.
        // The FFI function only reads from the pointer within the given length.
        let result = unsafe {
            ffi::proven_email_is_valid(bytes.as_ptr(), bytes.len())
        };
        core::bool_result_to_result(result)
    }
}
